### ISSUE TITLE
Subclassing DefaultViewService

### ISSUE THREAD

**Issue by depl0y** [IdentityServer/IdentityServer3#1549]
I would like to add some client specific urls to the login screen, but I don't want to go through the hassle of creating a completely new ViewService (Mainly, because I have already setup partial views that replace the current views.), so I thought I would subclass the DefaultViewService and extend a the login method.

```
public class MyViewService : DefaultViewService {

	public MyViewService(DefaultViewServiceOptions config, IViewLoader viewLoader) 
		: base(config, viewLoader) {
	}

	public override Task<System.IO.Stream> Login(Thinktecture.IdentityServer.Core.ViewModels.LoginViewModel model, Thinktecture.IdentityServer.Core.Models.SignInMessage message) {

		var links = new List<LoginPageLink>();

		links.Add(new LoginPageLink() {
			Href = "forgot-password?client=" + message.ClientId,
			Text = "Forgot password"
		});
		model.AdditionalLinks = links;

		return base.Login(model, message);
	}

}
```

After I created this class, I registered my subclassed ViewService like this:

```
factory.ViewService = new Registration<IViewService>(typeof(MyViewService));
```

But this doesn't work in my case, it crashes with the following exception message:

```
An error occurred when trying to create a controller of type 'ClientPermissionsController'. Make sure that the controller has a parameterless public constructor.
```

I am probably doing something wrong, but I would love some pointers that would show me how to add a link to the login screen.

So I was looking for something mentioned by @brockallen in https://github.com/IdentityServer/IdentityServer3/issues/1367

**Comment by brockallen** [comment]
Is there more to the stack track? I have a feeling your view service ctor has dependencies that might not be registered in the DI system. Check the docs for more info on how to do that.

**Comment by brockallen** [comment]
Oh yes, sorry - your ctor is right there. Ok, so to register a custom view service that derives from the default one, I'd suggest doing this instead:

```
factory.ViewService = new DefaultViewServiceRegistration<MyViewService>
```

**Comment by depl0y** [comment]
@brockallen Hmmm, I am unable to call the `DefaultViewServiceRegistration` like that (like a generic). Is this part of an extension library?

Because the code above generates the following compiler error for me:

```
'Thinktecture.IdentityServer.Core.Services.Default.DefaultViewServiceRegistration' cannot be used with type arguments.
```

I also can't find code that supports this in the sources of IS3

**Comment by brockallen** [comment]
It's in 2.0.0: https://www.nuget.org/packages/IdentityServer3

**Comment by depl0y** [comment]
@brockallen I performed an upgrade to 2.0.0-rc and it seems to work. I get my `AdditionalLinks` right where I want them. Too bad my layout is gone after the upgrade, probably changed around the paths for the partial HTML files?

Is there maybe some updated documentation about this?

**Comment by leastprivilege** [comment]
The folder name has changed. Use templates now.

**Comment by depl0y** [comment]
@leastprivilege Thanks. I figured that out last night when looking through the sources of v2.0.0.

Not sure how to add my EntityFrameworkServiceOptions though. Is this already available in 2.0?

**Comment by leastprivilege** [comment]
Yes - but that's a separate issue.
